File: script/Matlab/DynamicModel.m
function T = DynamicModel(acceleration, velocity, orientation)
    AUV_MASS               = 50.0;
    AUV_VOLUME             = 0.04949949;
    BUOYANCY_CENTER        = [0.0, 0.0, 0.0];
    X_SURFACE              = 0.73 * 0.17;
    Y_SURFACE              = X_SURFACE;
    Z_SURFACE              = X_SURFACE + Y_SURFACE;
    ROLL_SURFACE           = 0.0;
    PITCH_SURFACE          = 0.0;
    YAW_SURFACE            = 0.40 * 0.17;
    GRAVITY                = AUV_MASS * 9.81;
    BUOYANCY               = -AUV_VOLUME * 1025 * 9.81;
    resultingForce         = GRAVITY + BUOYANCY;

    ConstDragCoefficient   = [136.1473; 104.6571; 0.0; 0.0; 0.0; 76.4971];
    QuadDragCoefficient    = [3.7787; 51.7301; 0.0; 0.0; 0.0; -122.1888];
                    
    MassInertiaCoefficient = [[50, 0.0, 0.0, 0.0, 0.0, 0.0];
                             [0.0, 50, 0.0, 0.0, 0.0, 0.0];
                             [0.0, 0.0, 50, 0.0, 0.0, 0.0];
                             [0.0, 0.0, 0.0, 1.9, 0.0, 0.0];
                             [0.0, 0.0, 0.0, 0.0, 3.1, 0.0];
                             [0.0, 0.0, 0.0, 0.0, 0.0, 1.7]];
          
    M = MassInertiaCoefficient;    
    D = ComputeDamping(ConstDragCoefficient, QuadDragCoefficient, velocity);
    G = ComputeGravityVector(resultingForce, BUOYANCY, BUOYANCY_CENTER, orientation);

    T = M * acceleration + D .* velocity + G;
    
end

function GravityVectorT = ComputeGravityVector(resultingForce, Buoyancy, buoyancy_center, orientation)
        GravityVector(1) = resultingForce  * sin(orientation(2));
        GravityVector(2) = -resultingForce * sin(orientation(1)) * cos(orientation(2));
        GravityVector(3) = -resultingForce * cos(orientation(1)) * cos(orientation(2));

        GravityVector(4) = Buoyancy  * cos(orientation(2)) * (buoyancy_center(3) * sin(orientation(1)) - buoyancy_center(2) * cos(orientation(1)));
        GravityVector(5) = Buoyancy  * (buoyancy_center(1) * cos(orientation(1)) * cos(orientation(2)) - buoyancy_center(3) * sin(orientation(2)));
        GravityVector(6) = -Buoyancy * (buoyancy_center(1) * sin(orientation(1)) * cos(orientation(2)) - buoyancy_center(2) * sin(orientation(2)));
        GravityVectorT = [GravityVector(1);GravityVector(2);GravityVector(3);GravityVector(4);GravityVector(5);GravityVector(6)];
end

function Damping = ComputeDamping(ConstDragCoefficient, QuadDragCoefficient, Velocity)
    
    Damping = ConstDragCoefficient + QuadDragCoefficient .* (Velocity .* Velocity);
    
end

File: src/control_library/toolbox/DynamicModel.cc
/**
 * \file	DynamicModel.cc
 * \author	Olivier Lavoie <olavoie9507@gmail.com>
 *
 * \copyright Copyright (c) 2018 S.O.N.I.A. AUV All rights reserved.
 *
 * \section LICENSE
 *
 * This file is part of S.O.N.I.A. software.
 *
 * S.O.N.I.A. AUV software is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * S.O.N.I.A. AUV software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with S.O.N.I.A. AUV software. If not, see <http://www.gnu.org/licenses/>.
 */

#include "DynamicModel.h"


namespace control
{
    DynamicModel::DynamicModel(std::shared_ptr<DynamicModelParam> &auvDynamicParameters): auvDynamicParameters_(auvDynamicParameters) {}

    Eigen::VectorXd DynamicModel::Update(Eigen::VectorXd &acceleration, Eigen::VectorXd &velocity, Eigen::Vector3d &orientation)
    {
        Eigen::VectorXd damping = ComputeDamping(velocity).array() * velocity.array();
        Eigen::VectorXd wrench;

        wrench = auvDynamicParameters_->auvAddedMass * acceleration + damping + ComputeGravity(orientation);

        return wrench;
    }

    Eigen::VectorXd DynamicModel::ComputeDamping(Eigen::VectorXd &velocity)
    {
        return auvDynamicParameters_->auvDamping.constDragCoefficient.array() +  auvDynamicParameters_->auvDamping.quadDragCoefficient.array() * (velocity.array() * velocity.array());
    }

    Eigen::VectorXd DynamicModel::ComputeGravity(Eigen::Vector3d &orientation)
    {
        Eigen::VectorXd gravityVector = Eigen::VectorXd::Zero(CARTESIAN_SPACE);

        gravityVector[0] = auvDynamicParameters_->resultingForce  * sin(orientation[1]);
        gravityVector[1] = -auvDynamicParameters_->resultingForce * sin(orientation[0]) * cos(orientation[1]);
        gravityVector[2] = -auvDynamicParameters_->resultingForce * cos(orientation[0]) * cos(orientation[1]);

        gravityVector[3] = auvDynamicParameters_->buoyancyForce  * cos(orientation[1]) * (auvDynamicParameters_->auvBuoyancyCenter[2] * sin(orientation[0]) - auvDynamicParameters_->auvBuoyancyCenter[1] * cos(orientation[0]));
        gravityVector[4] = auvDynamicParameters_->buoyancyForce  * (auvDynamicParameters_->auvBuoyancyCenter[0] * cos(orientation[0]) * cos(orientation[1]) - auvDynamicParameters_->auvBuoyancyCenter[2] * sin(orientation[1]));
        gravityVector[5] = -auvDynamicParameters_->buoyancyForce * (auvDynamicParameters_->auvBuoyancyCenter[0] * sin(orientation[0]) * cos(orientation[1]) - auvDynamicParameters_->auvBuoyancyCenter[1] * sin(orientation[1]));
        return gravityVector;
    }

}